examples/keep-alive-framework/src/main/scala/com/mesosphere/usi/helloworld/KeepAliveFramework.scala

```diff
@@ -0,0 +1,124 @@
+import scala.concurrent.duration._
+
+class KeepAliveFramework(conf: Config) extends StrictLogging {
+
+  implicit val system = ActorSystem()
+  implicit val mat = ActorMaterializer()
+  implicit val ec = system.dispatcher
+
+  val settings = MesosClientSettings(conf.getString("master-url"))
+  val frameworkInfo = FrameworkInfo
+    .newBuilder()
+    .setUser(
+      new SystemProperties()
+        .get("user.name")
+        .getOrElse(throw new IllegalArgumentException("A local user is needed to launch Mesos tasks")))
+    .setName("KeepAliveFramework")
+    .addRoles("test")
+    .addCapabilities(FrameworkInfo.Capability.newBuilder().setType(FrameworkInfo.Capability.Type.MULTI_ROLE))
+    .setFailoverTimeout(0d)
+    .build()
+
+  val client: MesosClient = Await.result(MesosClient(settings, frameworkInfo).runWith(Sink.head), 10.seconds)
+
+  val runSpec = RunSpec(
+    resourceRequirements = List(ScalarRequirement(ResourceType.CPUS, 0.001), ScalarRequirement(ResourceType.MEM, 32)),
+    shellCommand = """echo "Hello, world" && sleep 5""",
+    role = "test"
+  )
+
+  def generatePodSpec() = {
+    val podId = PodId(s"hello-world.${UUID.randomUUID()}.1")
+
+    val podSpec = PodSpec(
+      id = podId,
+      goal = Goal.Running,
+      runSpec = runSpec
+    )
+
+    podSpec
+  }
+
+  val specsSnapshot = SpecsSnapshot(
+    podSpecs = (1 to 10).map(_ => generatePodSpec()),
+    reservationSpecs = Seq.empty
+  )
+
+  def createNewIncarnationId(podId: PodId): PodId = {
+    val idAndIncarnation = """^(.+\..*)\.(\d+)$""".r
+    val (podIdWithoutIncarnation, currentIncarnation) = podId.value match {
+      case idAndIncarnation(id, inc) =>
+        id -> inc.toLong
+    }
+    PodId(s"$podIdWithoutIncarnation.${currentIncarnation + 1}")
+  }
+
+  // KeepAliveWatcher looks for a terminal task even and then restarts the whole pod.
+  val keepAliveWatcher: Flow[StateEvent, SpecUpdated, NotUsed] = Flow[StateEvent].mapConcat {
+    // Main state event handler. We log happy events and restart the pod if something goes wrong
+    case s: StateSnapshot =>
+      logger.info(s"Initial state snapshot: $s")
+      Nil
+
+    case PodStatusUpdated(id, Some(PodStatus(_, taskStatuses))) =>
+      import TaskState._
+      def activeTask(status: TaskStatus) = Seq(TASK_STAGING, TASK_STARTING, TASK_RUNNING).contains(status.getState)
+
+      // We're only interested in the bad task statuses for our pod
+      val failedTasks = taskStatuses.filterNot { case (id, status) => activeTask(status) }
+
+      if (failedTasks.nonEmpty) {
+        logger.info(s"Restarting Pod $id")
+
+        val newId = createNewIncarnationId(id)
+
+        PodSpecUpdated(newId, Some(PodSpec(newId, Goal.Running, runSpec))) :: Nil
+      } else {
+        Nil
+      }
+
+    case e =>
+      logger.error(s"Unhandled event: $e") // we ignore everything else for now
+      Nil
+  }
+
+  val (snap, source, sink) = Scheduler.asSourceAndSink(specsSnapshot, client)
```

Comment:
This part is the core of the framework - please add some ASCII docs visualizing how it works and what happens inside!